contracts/eosio.bios/src/eosio.bios.cpp

```diff
@@ -17,6 +18,34 @@ void bios::setabi( name account, const std::vector<char>& abi ) {
    }
 }
 
+void bios::setfinalizer( const finalizer_policy& finalizer_policy ) {
+   require_auth( get_self() );
+   check(finalizer_policy.finalizers.size() > 0, "require at least one finalizer");
+
+   eosio::abi_finalizer_policy abi_finalizer_policy;
+   abi_finalizer_policy.fthreshold = finalizer_policy.threshold;
+   abi_finalizer_policy.finalizers.reserve(finalizer_policy.finalizers.size());
+
+   const std::string pk_prefix = "PUB_BLS";
+   const std::string sig_prefix = "SIG_BLS";
+
+   for (const auto& f: finalizer_policy.finalizers) {
+      // basic key format checks
+      check(f.public_key.substr(0, pk_prefix.length()) == pk_prefix, "public key not started with PUB_BLS");
+      check(f.pop.substr(0, sig_prefix.length()) == sig_prefix, "proof of possession signature not started with SIG_BLS");
+
+      // proof of possession of private key check
+      const auto pk = eosio::decode_bls_public_key_to_g1(f.public_key);
+      const auto signature = eosio::decode_bls_signature_to_g2(f.pop);
+      check(eosio::bls_pop_verify(pk, signature), "proof of possession failed");
+
+      std::vector<char> pk_vector(pk.begin(), pk.end());
+      abi_finalizer_policy.finalizers.emplace_back(eosio::abi_finalizer_authority{f.description, f.weight, std::move(pk_vector)});
```

Comment:
The action should check that a duplicate key is not being added.

Reply:
Additionally, other checks should be added to ensure the `set_finalizers` host function call does not fail in Leap:

* The number of finalizers in the proposal must be no greater than `config::max_finalizers`.
* Each `description` must have a length no greater than `config::max_finalizer_description_size`.
* The sum of the weights must not overflow `uint64_t`.
* The threshold must be greater than half the sum of the weights.
* There should not be any duplicates in the list of public keys proposed in the finalizer policy.

Reply:
Thanks! All additional checks and new tests are done by https://github.com/AntelopeIO/reference-contracts/pull/33/commits/312f46adff04e1f2a421a3b4b4d0219346913132 and https://github.com/AntelopeIO/reference-contracts/pull/33/commits/d5bf2414283ce902c3d13b06bbc823ddb3b2437c